File: app/src/main/java/com/aero51/moviedatabase/viewmodel/TvShowsViewModel.java
package com.aero51.moviedatabase.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.paging.PagedList;

import com.aero51.moviedatabase.repository.TvShowsRepository;
import com.aero51.moviedatabase.repository.model.NetworkState;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.AiringTvShowsPage;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.PopularTvShowsPage;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TrendingTvShowsPage;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowGenresResponse;
import com.aero51.moviedatabase.utils.AppExecutors;
import com.aero51.moviedatabase.utils.Resource;

import java.util.List;

public class TvShowsViewModel extends AndroidViewModel {
    private TvShowsRepository tvShowsRepository;

    public TvShowsViewModel(@NonNull Application application) {
        super(application);
        AppExecutors executors = new AppExecutors();
        tvShowsRepository=new TvShowsRepository(application, executors);
    }
    public LiveData<PagedList<PopularTvShowsPage.PopularTvShow>> getTvPopularResultsPagedList() {
        return tvShowsRepository.getTvPopularResultsPagedList();
    }

    //used to get the page number by adding observer
    public LiveData<PopularTvShowsPage> getPopularLiveTvShowPage() {
        return tvShowsRepository.getLastPopularTvShowPage();
    }

    public LiveData<NetworkState> getPopularTvShowsNetworkState() {
        return tvShowsRepository.getPopularNetworkState();
    }


    public LiveData<PagedList<AiringTvShowsPage.AiringTvShow>> getTvAiringResultsPagedList() {
        return tvShowsRepository.getTvAiringResultsPagedList();
    }

    //used to get the page number by adding observer
    public LiveData<AiringTvShowsPage> getAiringLiveTvShowPage() {
        return tvShowsRepository.getLastAiringTvShowPage();
    }

    public LiveData<NetworkState> getAiringTvShowsNetworkState() {
        return tvShowsRepository.getAiringNetworkState();
    }

    public LiveData<PagedList<TrendingTvShowsPage.TrendingTvShow>> getTvTrendingResultsPagedList() {
        return tvShowsRepository.getTvTrendingResultsPagedList();
    }

    //used to get the page number by adding observer
    public LiveData<TrendingTvShowsPage> getTrendingLiveTvShowPage() {
        return tvShowsRepository.getLastTrendingTvShowPage();
    }

    public LiveData<NetworkState> getTrendingTvShowsNetworkState() {
        return tvShowsRepository.getTrendingNetworkState();
    }
    public LiveData<Resource<List<TvShowGenresResponse.TvShowGenre>>> getTvShowsGenres() {
        return  tvShowsRepository.loadTvShowsGenres();
    }

}

File: app/src/main/java/com/aero51/moviedatabase/viewmodel/TvShowsViewModel.kt
package com.aero51.moviedatabase.viewmodel

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import androidx.paging.PagedList
import com.aero51.moviedatabase.repository.TvShowsRepository
import com.aero51.moviedatabase.repository.model.NetworkState
import com.aero51.moviedatabase.repository.model.tmdb.movie.MoviesByGenrePage
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.*
import com.aero51.moviedatabase.utils.AppExecutors
import com.aero51.moviedatabase.utils.Constants
import com.aero51.moviedatabase.utils.Resource
import kotlinx.coroutines.launch

class TvShowsViewModel(application: Application) : AndroidViewModel(application) {
    private val tvShowsRepository: TvShowsRepository

    init {
        val executors = AppExecutors()
        tvShowsRepository = TvShowsRepository(application, executors)
    }

    val tvPopularResultsPagedList: LiveData<PagedList<PopularTvShowsPage.PopularTvShow>>?
        get() = tvShowsRepository.tvPopularResultsPagedList

    //used to get the page number by adding observer
    val popularLiveTvShowPage: LiveData<PopularTvShowsPage>
        get() = tvShowsRepository.lastPopularTvShowPage

    val popularTvShowsNetworkState: LiveData<NetworkState>
        get() = tvShowsRepository.popularNetworkState

    val tvAiringResultsPagedList: LiveData<PagedList<AiringTvShowsPage.AiringTvShow>>?
        get() = tvShowsRepository.tvAiringResultsPagedList

    //used to get the page number by adding observer
    val airingLiveTvShowPage: LiveData<AiringTvShowsPage>
        get() = tvShowsRepository.lastAiringTvShowPage

    val airingTvShowsNetworkState: LiveData<NetworkState>
        get() = tvShowsRepository.airingNetworkState

    val tvTrendingResultsPagedList: LiveData<PagedList<TrendingTvShowsPage.TrendingTvShow>>?
        get() = tvShowsRepository.tvTrendingResultsPagedList

    //used to get the page number by adding observer
    val trendingLiveTvShowPage: LiveData<TrendingTvShowsPage>
        get() = tvShowsRepository.lastTrendingTvShowPage

    val trendingTvShowsNetworkState: LiveData<NetworkState>
        get() = tvShowsRepository.trendingNetworkState

    val tvShowsGenres: LiveData<Resource<List<TvShowGenresResponse.TvShowGenre>>>
        get() = tvShowsRepository.loadTvShowsGenres()

    //used to get the page number by adding observer
    val tvShowsByGenrePage: LiveData<TvShowsByGenrePage>
        get() = tvShowsRepository.lastTvShowsByGenrePage

    fun getTvShowsByGenre(genreId: Int):  LiveData<PagedList<TvShowsByGenrePage.TvShowByGenre>>?{
        return tvShowsRepository.loadTvShowsByGenre(genreId)
    }

    fun  tvShowsByGenreDataValidationCheck(genreId: Int){
        viewModelScope.launch {
            // suspend and resume make this database request main-safe
            // so our ViewModel doesn't need to worry about threading
            tvShowsRepository.checkIfTvShowsByGenreNeedsRefresh(genreId)
        }
    }
    fun  popularShowsDataValidationCheck(){
        viewModelScope.launch {
            // suspend and resume make this database request main-safe
            // so our ViewModel doesn't need to worry about threading
            tvShowsRepository.checkIfPopularTvShowsNeedsRefresh()
        }
    }
    fun  trendingShowsDataValidationCheck(){
        viewModelScope.launch {
            // suspend and resume make this database request main-safe
            // so our ViewModel doesn't need to worry about threading
            tvShowsRepository.checkIfTrendingTvShowsNeedsRefresh()
        }
    }

    fun  airingShowsDataValidationCheck(){
        viewModelScope.launch {
            // suspend and resume make this database request main-safe
            // so our ViewModel doesn't need to worry about threading
            tvShowsRepository.checkIfAiringTvShowsNeedsRefresh()
        }
    }

}